components/raftstore/src/store/fsm/apply.rs

```diff
@@ -2981,6 +3045,36 @@ where
             ));
         }
 
+        let voter_replicated_index = req.get_compact_log().get_voter_replicated_index();
+        // If there is any voter lagging behind, the log truncation of the witness
+        // shouldn't be triggered even if it's force mode(raft log size/count exceeds
+        // the threshold or raft engine purge), otherwise the witness can't help the
+        // lagging voter catch up logs when leader is down. In this situation Compact
+        // index should be queued. If witness receives a voter_replicated_index
+        // that is larger than the pending compact index, logs can be deleted.
+        if self.peer.is_witness && voter_replicated_index < compact_index {
+            match self.pending_cmds.pop_compact(voter_replicated_index) {
+                Some(cmd) => {
```

Comment:
The current compact task should be pushed in queue even there are pending compacts. A case seems to be missing.